ee/control/bucket_consumer.go

```diff
@@ -0,0 +1,58 @@
+package control
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"go.etcd.io/bbolt"
+)
+
+// BucketConsumer processes control server updates for a named bucket
+type BucketConsumer struct {
+	logger     log.Logger
+	db         *bbolt.DB
+	bucketName string
+}
+
+func NewBucketConsumer(logger log.Logger, db *bbolt.DB, bucketName string) *BucketConsumer {
+	bc := &BucketConsumer{
+		logger:     logger,
+		db:         db,
+		bucketName: bucketName,
+	}
+
+	return bc
+}
+
+func (bc *BucketConsumer) Update(data io.Reader) {
+	var kvPairs map[string]string
+	if err := json.NewDecoder(data).Decode(&kvPairs); err != nil {
+		level.Error(bc.logger).Log(
+			"msg", "failed to decode bucket consumer json",
+			"bucketName", bc.bucketName,
+			"data", data,
+			"err", err,
+		)
+	}
+
+	bc.db.Update(func(tx *bbolt.Tx) error {
+		// Clear the bucket first
+		tx.DeleteBucket([]byte(bc.bucketName))
+
+		bucket, err := tx.CreateBucketIfNotExists([]byte(bc.bucketName))
```

Comment:
There is some additional complexity within `DeleteBucket` - it will handle deleting child buckets, for example. I couldn't find any API for "remove all keys from the bucket, and leave the bucket empty". So the alternate approach is iterating through with `ForEach` and calling `Delete` for each key.

Deleting the bucket does seem a bit heavy-handed, but I also don't think this particular operation would get executed very frequently, so I'm not too concerned.

`DeleteBucket` does return an error - I ignored it because it's expected the first time this happens the bucket won't exist. I could log the error and proceed, though?

Also - I assumed we wanted to blow away the bucket data and replace it entirely with the fresh payload, but is that necessarily correct? I suppose I don't know exactly all the ways this table is intended to be used. I went with this approach to avoid a scenario where the table grows in size forever without ever being purged. But I could also envision use cases where we don't want to lose what has been added to the table in previous updates, and it may be tedious to ensure the newest payload contains all previous key-values we want to persist.

If we want to support both, the update payload could optionally contain a boolean indicating whether or not we want to clear all previous bucket/table data.

Reply:
I revamped this area a bit to avoid bucket deletion and provide a means to delete keys by sending along empty values.

Reply:
Took your suggestion and now pruning stale keys not in the latest update.